### PR TITLE
Add Imagick CMYK to SRGB conversion

### LINKED ISSUE #22375: CMYK images are rendered with inverted colors
### Preconditions (*)
1. Magento 2.3.1
2. ImageMagick as image processor
3. One CMYK image

### Steps to reproduce (*)
1. Install Magento 2.3.1 and use ImageMagick as image processor. 
2. Create a product and upload an image in CMYK color profile. 
3. The image displays properly in backend in product edit page because it shows the original image.
4. Check the product in frontend and notice that the image is resized via ImageMagick but the colors are inverted. 

### Expected result (*)
1. The image displays properly would be ideal but at least the user should be warned that the image is saved in a color space that Magento 2 cannot handle so it will not be be converted correctly.

### Actual result (*)
1. Image is resized but the colors are inverted. 
image

Unfortunately the users of an eshop are not usually proficient in the handling of apps like photoshop and they will probably upload lots of images they downloaded from the web or from their providers. Thus it's very possible that they will encounter situations like this. 

I think the system should at least warn them that it cannot process the image correctly if it can't handle the conversion.
Perhaps Imagick::transformImageColorspace could be used to handle it silently? 
 
Here's a sample image
madrid-orange_1_1

### PR DESCRIPTION
**Pull request by tdgroot**
Add Imagick CMYK to SRGB conversion

### Description (*)
When Imagick adapter detects that the opened image has a CMYK colorspace, convert the colorspace to SRGB.

### Related Pull Requests

1. magento/magento2#25277

### Fixed Issues (if relevant)

1. Fixes magento/magento2#22375

### Manual testing scenarios (*)

1. Set Imagick as the image adapter
2. Create a product and configure the following file: 
image
3. The colors on the frontend of the resized image should be correct.

### Questions or comments
Not sure if the CMYK image added to the unit test has any copyrights to it. @ioweb-gr Can you confirm if the image can be used for this purpose?

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by krisdante** [comment]
The test is here, that's a fact. But the test is not testing what this problem is all about.
IMO the test shall be extended and do the color probe of  some pixel, and check that this color probe in actually orangish not blueish

**Comment by tdgroot** [comment]
> The test is here, that's a fact. But the test is not testing what this problem is all about.
> IMO the test shall be extended and do the color probe of some pixel, and check that this color probe in actually orangish not blueish

I do not completely agree with you. While it would be very thorough to test if the image pixels are corrected, this is not completely Magento territory. That is, because Magento should not test if Imagick is doing its work.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 1]
```diff
@@ -35,6 +38,19 @@ class ImageMagick extends AbstractAdapter
         'sharpen' => ['radius' => 4, 'deviation' => 1],
     ];
 
+    /**
+     * Colorspace of the image
+     *
+     * @var int
+     */
+    protected $colorspace = -1;
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Please, make property private

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 2]
```diff
@@ -35,6 +38,19 @@ class ImageMagick extends AbstractAdapter
...
+    /**
+     * Original colorspace of the image
+     *
+     * @var int
+     */
+    protected $originalColorspace = -1;
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Please, make property private

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 3]
```diff
@@ -562,4 +580,42 @@ private function addSingleWatermark($positionX, int $positionY, \Imagick $waterm
             $compositeChannels
         );
     }
+
+    /**
+     * Get and store the image colorspace.
+     *
+     * @return int
+     */
+    public function getColorspace(): int
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Please, make method private

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 4]
```diff
@@ -562,4 +580,42 @@ private function addSingleWatermark($positionX, int $positionY, \Imagick $waterm
...
+
+        return $this->colorspace;
+    }
+
+    /**
+     * Get the original image colorspace.
+     *
+     * @return int
+     */
+    public function getOriginalColorspace(): int
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Please, make method private

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 5]
```diff
@@ -562,4 +580,42 @@ private function addSingleWatermark($positionX, int $positionY, \Imagick $waterm
...
+
+    /**
+     * Get and store the image colorspace.
+     *
+     * @return int
+     */
+    public function getColorspace(): int
+    {
+        if ($this->colorspace === -1) {
+            $this->originalColorspace = $this->colorspace = $this->_imageHandler->getImageColorspace();
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: Please, make an assignment in the separate lines for better readability

**Review by sidolov — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 3 above)
Comment (tdgroot) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: @sidolov I can do that, but then this behavior cannot be tested like it is right now. It depends on these functions being public.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 4 above)
Comment (tdgroot) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: @sidolov I can do that, but then this behavior cannot be tested like it is right now. It depends on these functions being public.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 3 above)
Comment (sidolov) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: I understand that but it's better to eliminate the possibility of inheritance here.

Comment (krisdante) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:592]: Those tests are not very useful anyway. The "business" value it to have proper colors, not to get palette code properly

**Comment by gabrieldagama** [comment]
The risk was set to`medium` due to: the suggested changes are made on the framework and may affect other areas or cause undesired behavior.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php [hunk 6]
```diff
@@ -35,6 +36,24 @@ class ImageMagick extends AbstractAdapter
         'sharpen' => ['radius' => 4, 'deviation' => 1],
     ];
 
+    /**
+     * @var \Imagick
+     */
+    protected $_imageHandler;
```
Comment (sidolov) [inline lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:42]: Why you need this var? It should be inherited from the parent class

**Review by sidolov — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 6 above)
Comment (tdgroot) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php:42]: It provides type information for within the ImageMagick class.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 3 above)
Comment (tdgroot) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: This method is private now.

File: lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php (hunk 4 above)
Comment (tdgroot) [reply lib/internal/Magento/Framework/Image/Adapter/ImageMagick.php]: This method was need for testing, but I deleted now in favor of requested changes.

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Comment by engcom-Golf** [comment]
@magento run all tests

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-8296](https://jira.corp.magento.com/browse/ENGCOM-8296) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Bravo** [comment]
:heavy_check_mark: QA Passed

### Manual testing scenario
1. Install _ImageMagick_ image processor + php extension installed (`sudo apt install php-imagick`)
2. From Magento Admin, go to **Stores - Configuration - Advanced - Developer**, expand, **Image Processing Settings**
3. Set **Image Adapter** to _ImageMagick_. 
Selection_153
**Save Config** and clear Magento Cache
4. Go to ***Catalog - Products*, click **Add Product**
5. Fill in all required fields, set _Category_
6. Expand **Images And Videos** and upload a **CMYK color** image
ima_vid_sett
7. **Save** the product and assert that the color is original, not inverted
8. Go to Storefront and check the previously created products image

**BEFORE** applying the changes provided in this PR, the color of CMYK images was inverted
thumb
prod_page
Peek 2020-10-02 11-01

**AFTER** switching to this PR, the newly added CMYK Product images are displayed correctly on Storefront
after1
after2

**Pull request merged**